src/qibolab/compilers/compiler.py

```diff
@@ -0,0 +1,143 @@
+
+    def __delitem__(self, item):
+        """Remove rule for the given gate."""
+        if item not in self.rules:
+            raise_error(KeyError, f"Cannot remove {item} from compiler because it does not exist.")
+        del self.rules[item]
+
+    def register(self, gate_cls):
+        """Decorator for registering a function as a rule in the compiler.
+
+        Using this decorator is optional. Alternatively the user can set the rules directly
+        via ``__setitem__`.
+
+        Args:
+            gate_cls: Qibo gate object that the rule will be assigned to.
+        """
+
+        def inner(func):
+            self[gate_cls] = func
+            return func
+
+        return inner
+
+    def compile(self, circuit, platform):
+        """Transforms a circuit to pulse sequence.
+
+        Args:
+            circuit (qibo.models.Circuit): Qibo circuit that respects the platform's
+                connectivity and native gates.
+            platform (qibolab.platforms.abstract.AbstractPlatform): Platform used
+                to load the native pulse representations.
+
+        Returns:
+            sequence (qibolab.pulses.PulseSequence): Pulse sequence that implements the circuit.
+            measurement_map (dict): Map from each measurement gate to the sequence of  readout pulses
+                implementing it.
+        """
+        sequence = PulseSequence()
+        # FIXME: This will not work with qubits that have string names
+        # TODO: Implement a mapping between circuit qubit ids and platform ``Qubit``s
+        virtual_z_phases = defaultdict(int)
+
+        measurement_map = {}
+        # keep track of gates that were already added to avoid adding them twice
+        already_processed = set()
+        # process circuit gates
+        for moment in circuit.queue.moments:
+            moment_start = sequence.finish
+            for gate in moment:
+                if gate is None or gate in already_processed:
+                    continue
+
+                rule = self[gate.__class__]
+                # get local sequence and phases for the current gate
+                gate_sequence, gate_phases = rule(gate, platform)
+
+                # update global pulse sequence
+                # determine the right start time based on the availability of the qubits involved
+                all_qubits = {*gate_sequence.qubits, *gate.qubits}
+                start = max(sequence.get_qubit_pulses(*all_qubits).finish, moment_start)
+                # shift start time and phase according to the global sequence
+                for pulse in gate_sequence:
+                    pulse.start += start
+                    if not isinstance(pulse, ReadoutPulse):
+                        pulse.relative_phase += virtual_z_phases[pulse.qubit]
+                    sequence.add(pulse)
+
+                # update virtual Z phases
+                for qubit, phase in gate_phases.items():
+                    virtual_z_phases[qubit] += phase
```

Comment:
This part "only" depends on `gate, platform, sequence, virtual_z_phases`, so you could split into its own function.

Indeed, if you do not modify `sequence` inside (postponing the `sequence.add` call), but instead return a series of pulses (the `gate_sequence`) and you collect them in the outer loop (something like `gate_sequences`), you can modify `sequence` at the very end of the outer iteration, and use `sequence.finish` instead of `moment_start`.

Reply:
I moved this to another function.

> Indeed, if you do not modify `sequence` inside (postponing the `sequence.add` call), but instead return a series of pulses (the `gate_sequence`) and you collect them in the outer loop (something like `gate_sequences`), you can modify `sequence` at the very end of the outer iteration, and use `sequence.finish` instead of `moment_start`.

Maybe we could have a `moment_sequence` for each moment and add it to the global `sequence` at the end of the outer iteration (over moments), however I am not sure to what extend this would simplify this code, which is already not very complicated.

Another simple way to reduce the arguments of the new function (and also slightly simplify other parts of the code) is to move `virtual_z_phases` inside the `PulseSequence` object. I think we considered this in the past, but I don't remember if there were any strong arguments against.

Reply:
> Maybe we could have a `moment_sequence` for each moment and add it to the global `sequence` at the end of the outer iteration (over moments), however I am not sure to what extend this would simplify this code, which is already not very complicated.

The `gate_sequences` I mentioned should be exactly your `moment_sequence`.

> Another simple way to reduce the arguments of the new function (and also slightly simplify other parts of the code) is to move `virtual_z_phases` inside the `PulseSequence` object. I think we considered this in the past, but I don't remember if there were any strong arguments against.

The idea is, if possible, to share less "global" state (so make things as local as possible). Alternatively, when it is not possible to reduce the number of shared structures, if they are shared in the same place there might be a value in packaging them in a box. If `PulseSequence` can act as such a box, it might be good enough, since that is already shared.